kernel: Add k_thread_runtime_stats_is_enabled function

rruuaanng opened the pull request:
Add 'k_thread_runtime_stats_is_enabled' to 'usage.c', which's used to check whether runtime statistics collection is enabled for a thread.

It can determine whether the thread has enabled runtime statistics collection when the program is running. It allows users to add conditional statements based on this function when using it.

for example:
```
void f()
{
    // Assume there is a thread id
    if (k_thread_runtime_stats_is_enabled(id)) {
        /* do something */
    }
    ...
}
```
And it can provide a complete support when similar operations are needed. for example `enable`, `disable`, `is_enabled`.
For example, its use in the implementation of CLOCK_THREAD_CPUTIME_ID actually lacks k_thread_runtime_stats_is_enabled, which makes the entire function impossible to implement safely. It is necessary.

..

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -6224,6 +6224,17 @@ int k_thread_runtime_stats_enable(k_tid_t thread);
+ * @param thread ID of thread
+ * @return -EINVAL if invalid thread ID, otherwise return usage_lock status
+ */
+int k_thread_runtime_stats_is_enabled(k_tid_t thread);
```
Is there a specific reason that this shouldn't be inline or just a compile-time check?
```suggestion
static inline int k_thread_runtime_stats_is_enabled(void)
{
    if (IS_ENABLED(CONFIG_THREAD_RUNTIME_STATS)) {
        return true;
    }

    return false;
}
```

rruuaanng replied:
It seems that you are right, but I think that users can choose to turn it off or on during subsequent development, that is, it should be able to obtain the result dynamically at runtime rather than through macros determined at compile time.

cfriedt reviewed (commented):
Thanks for all of the POSIX clock work.

Just a question / suggestion regarding runtime function call overhead.

cfriedt commented on lib/posix/options/clock.c:
```diff
@@ -75,6 +93,10 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 		(void)__posix_clock_get_base(clock_id, &base);
 		break;
 
+	case CLOCK_THREAD_CPUTIME_ID:
+		(void)__posix_clock_get_base(clock_id, &base);
+		break;
+
```
```suggestion
#ifdef CONFIG_POSIX_THREAD_CPUTIME
	case CLOCK_THREAD_CPUTIME_ID:
		(void)__posix_clock_get_base(clock_id, &base);
		break;
#endif
```

cfriedt commented on lib/posix/options/clock.c:
```diff
@@ -43,6 +43,24 @@ int z_impl___posix_clock_get_base(clockid_t clock_id, struct timespec *base)
 		}
 		break;
+#ifdef CONFIG_SCHED_THREAD_USAGE_ANALYSIS
+	case CLOCK_THREAD_CPUTIME_ID:
```
```suggestion
#ifdef CONFIG_POSIX_THREAD_CPUTIME
	case CLOCK_THREAD_CPUTIME_ID:
```

rruuaanng replied:
It seems that `CLOCK_THREAD_CPUTIME_ID` is defined directly in `zephyr/posix/time.h`

cfriedt commented on lib/posix/options/clock.c:
```diff
@@ -43,6 +43,24 @@ int z_impl___posix_clock_get_base(clockid_t clock_id, struct timespec *base)
+			base->tv_nsec = ns % NSEC_PER_SEC;
+		}
+		break;
+#endif /* CONFIG_SCHED_THREAD_USAGE_ANALYSIS */
```
```suggestion
#endif
#endif /* CONFIG_SCHED_THREAD_USAGE_ANALYSIS */
```

cfriedt reviewed (changes requested):
Just some suggestions around `CONFIG_POSIX_THREAD_CPUTIME`

rruuaanng commented:
I've made change, Please riewer again!

ycsin commented on lib/posix/options/clock.c:
```diff
@@ -43,6 +43,24 @@ int z_impl___posix_clock_get_base(clockid_t clock_id, struct timespec *base)
+			base->tv_nsec = ns % NSEC_PER_SEC;
+		}
+		break;
+#endif /* CONFIG_SCHED_THREAD_USAGE_ANALYSIS */
```
nit: this doesn't match L46 & L47

rruuaanng replied:
Yes, I noticed it. But I'm wondering if I should change it to `CONFIG_SCHED_THREAD_USAGE_ANALYSIS && CONFIG_POSIX_THREAD_CPUTIME`, or ignore it.

ycsin replied:
I'd change it to `CONFIG_SCHED_THREAD_USAGE_ANALYSIS && CONFIG_POSIX_THREAD_CPUTIME`

rruuaanng replied:
I've change, Please to see :)

peter-mitsis reviewed (dismissed):
Kernel side seems reasonable to me.

stephanosio commented on kernel/usage.c:
```diff
@@ -272,6 +272,22 @@ int k_thread_runtime_stats_disable(k_tid_t  thread)
 	return 0;
 }
+
+int k_thread_runtime_stats_is_enabled(k_tid_t thread)
```
```suggestion
bool k_thread_runtime_stats_is_enabled(k_tid_t thread)
```

Please use `bool` for boolean data types.

stephanosio commented on lib/posix/options/clock.c:
```diff
@@ -43,6 +43,24 @@ int z_impl___posix_clock_get_base(clockid_t clock_id, struct timespec *base)
+			k_thread_runtime_stats_t stats;
+			k_tid_t thread = k_current_get();
+
+			if (!k_thread_runtime_stats_is_enabled(thread)) {
+				errno = EINVAL;
+				return -1;
+			}
```
You are evaluating `int` as boolean data type here. Please make up your mind on whether you want this function to return boolean data or something else ...

stephanosio commented on include/zephyr/kernel.h:
```diff
@@ -6224,6 +6224,17 @@ int k_thread_runtime_stats_enable(k_tid_t thread);
+ * specified thread and returns the current usage_lock status.
+ *
+ * @param thread ID of thread
+ * @return -EINVAL if invalid thread ID, otherwise return usage_lock status
+ */
+int k_thread_runtime_stats_is_enabled(k_tid_t thread);
```
This is a bad design.

If you want this function to return a "status," then make it return a status (i.e. `0` on success) and nothing else -- `track_usage` should be set via a `bool *` parameter.

If you want to make this function return `bool track_usage`, then make it return `bool`.

rruuaanng replied:
I don't seem to fully understand. If you're referring to the fact that it shouldn't mix error handling, which leads to a return value range of [0, 1, -1], my thought is that error handling can make it easier to catch errors when using this function. If we assume its implementation is safe, meaning tid cannot be None, I could remove the error handling, making the return value range [0, 1].

rruuaanng commented:
It seems that this test fails.
```
Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/ipi_optimize/src/main.c:386: ipi_test_thread_priority_set_lower: (set[i] == 1 is false)
CPU0 got 0 IPIs.

FAIL - test_thread_priority_set_lower in 0.041 seconds
```
This doesn't seem to be related to my changes, the difference is only changing int to bool. Hmmm....

cfriedt commented on tests/posix/common/src/clock.c:
Please also add entries for `CLOCK_PROCESS_CPUTIME_ID`  and `CLOCK_THREAD_CPUTIME_ID` to `clocks[]`,  `settable[]`, and `args[]`, as needed.

https://github.com/zephyrproject-rtos/zephyr/blob/193eeaef0c04570b80a954b58ccd98075150fb11/tests/posix/common/src/clock.c#L22

https://github.com/zephyrproject-rtos/zephyr/blob/193eeaef0c04570b80a954b58ccd98075150fb11/tests/posix/common/src/clock.c#L245

rruuaanng replied:
I've made change :)

cfriedt replied:
@rruuaang - it looks like the additional entry is missing for `settable[]`.

rruuaanng replied:
I was looking for it and changed it, but it seems that adding it to the clock gives an error in CI :(

rruuaanng commented on tests/posix/common/src/clock.c:
```diff
@@ -22,6 +22,9 @@ static const struct timespec ref_ts = {1514821501, NSEC_PER_SEC / 2U};
 	CLOCK_MONOTONIC,
 	CLOCK_REALTIME,
+#ifdef CONFIG_POSIX_THREAD_CPUTIME
+	CLOCK_THREAD_CPUTIME_ID
```
I've add it.

cfriedt replied:
Please add a `settable[]` entry, and also one for `args[]` down below (a valid input).

cfriedt commented on tests/posix/common/src/clock.c:
```diff
@@ -105,6 +108,7 @@ ZTEST(clock, test_gettimeofday)
 	/* Validate gettimeofday API */
 	zassert_ok(gettimeofday(&tv, NULL));
 	zassert_ok(clock_gettime(CLOCK_REALTIME, &rts));
+	zassert_ok(clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts));
```
Likely this should have an ifdef around it.

rruuaanng commented:
I've made changes :), CI look's fine.

ycsin reviewed (changes requested):
LGTM, minor issues

cfriedt reviewed (changes requested):
Probably unnecessary to take the spinlock in the case of gathering execution cycles

cfriedt commented on lib/posix/options/clock.c:
```diff
@@ -75,6 +93,12 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
+#ifdef CONFIG_POSIX_THREAD_CPUTIME
+	case CLOCK_THREAD_CPUTIME_ID:
+		(void)__posix_clock_get_base(clock_id, &base);
+		break;
```
In the case of execution cycles, I think we don't want to add anything on top of `base`.


```suggestion
   *ts = base;

		return 0;
```

rruuaanng replied:
I agree, it seems that we only need to get the execution cycle, and no other content is needed.

cfriedt reviewed (changes requested):
In the case of execution cycles, I think we don't want to add anything on top of base.

rruuaanng commented on lib/posix/options/clock.c:
```diff
@@ -43,6 +43,24 @@ int z_impl___posix_clock_get_base(clockid_t clock_id, struct timespec *base)
 		break;
+#if defined(CONFIG_SCHED_THREAD_USAGE_ANALYSIS) && defined(CONFIG_POSIX_THREAD_CPUTIME)
+	case CLOCK_THREAD_CPUTIME_ID:
+		K_SPINLOCK(&rt_clock_base_lock) {
```
This is just to make it safer in SMP environment. Look this
```
case CLOCK_REALTIME:
K_SPINLOCK(&rt_clock_base_lock) {
	*base = rt_clock_base;
}
break;
```

cfriedt replied:
Right - spinlocks are used for synchronization with SMP enabled. However, the code does not read or write the `rt_clock_base` variable, which the spinlock protects (nor should it, since it's measuring per-thread execution time, which is not relative to `CLOCK_REALTIME`).

rruuaanng replied:
I've remove it :)

rruuaanng commented:
I've made changes :)

cfriedt commented on lib/posix/options/clock.c:
```diff
@@ -43,6 +43,22 @@ int z_impl___posix_clock_get_base(clockid_t clock_id, struct timespec *base)
 			*base = rt_clock_base;
 		}
 		break;
+#if defined(CONFIG_SCHED_THREAD_USAGE_ANALYSIS) && defined(CONFIG_POSIX_THREAD_CPUTIME)
```
```suggestion
#ifdef CONFIG_POSIX_THREAD_CPUTIME
```

ycsin replied:
the `config POSIX_THREAD_CPUTIME` probably has to be updated for the dependencies

cfriedt replied:
That should get rid of the following error

```shell
START - test_clock_gettime

    Assertion failed at WEST_TOPDIR/zephyr/tests/posix/common/src/clock.c:94: clock_test_clock_gettime: ts.tv_nsec equal to -1

 FAIL - test_clock_gettime in 0.001 seconds
```

rruuaanng replied:
After completing the push, maybe we can look forward to CI :-------)

cfriedt replied:
Technically, `twister` should be run locally, prior to pushing, so that we have more confidence that CI won't break.

rruuaanng replied:
What confuses me is that it runs fine locally :(

ycsin commented:
> It seems that removing the spinlock does not pass CI, which is what I expected. Maybe the implementation after removing it is not that reliable.



> I've made changes :)
> 
> Edit Hmmm.... It seems that removing the spinlock does not pass CI, which is what I expected. Maybe the implementation after removing it is not that reliable.

The test fails on `qemu_cortex_a9_xc7z007s` which doesn't seem like a SMP board/soc, I suspect something's up with the unexpected code path due to Kconfig, could you connect a debugger to it and check what actually happened?

cfriedt commented on tests/kernel/usage/thread_runtime_stats/src/test_thread_runtime_stats.c:
```diff
@@ -225,6 +225,7 @@ ZTEST(usage_api, test_thread_stats_enable_disable)
 	k_sleep(K_TICKS(2));
 
 	k_thread_runtime_stats_enable(tid);
+	zassert_true(k_thread_runtime_stats_is_enabled(tid));
```
Above this, on line 215, can you also please add 

```cpp	zassert_false(k_thread_runtime_stats_is_enabled(tid));
```

cfriedt reviewed (changes requested):
Just one minor change for kernel test code.

However, additionally, can you please ensure that code commits and test commits are separate?

Specifically, in this order:

1. kernel code changes (incl header)
2. kernel test changes
3. POSIX code changes
4. POSIX test changes

Sorry to make you jump through yet another hoop

rruuaanng commented:
Of course.

My personal understanding of this is that the test should be submitted with the new function, and make sure that they can be synchronized when rolling back later.

cfriedt commented:
> My personal understanding of this is that the test should be submitted with the new function, and make sure that they can be synchronized when rolling back later.

That's one way to do it, but in Zephyr we almost always separate code and test commits.

stephanosio commented on tests/posix/common/src/clock.c:
```diff
@@ -105,6 +111,9 @@ ZTEST(clock, test_gettimeofday)
 	/* Validate gettimeofday API */
 	zassert_ok(gettimeofday(&tv, NULL));
 	zassert_ok(clock_gettime(CLOCK_REALTIME, &rts));
+#ifdef CONFIG_POSIX_THREAD_CPUTIME
+	zassert_ok(clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts));
+#endif /* CONFIG_POSIX_THREAD_CPUTIME */
```
Why are we testing `clock_gettime()` inside `test_gettimeofday`, which is supposed to test `gettimeofday()`?

cfriedt replied:
Good catch as well. This bit can be removed

stephanosio commented on include/zephyr/kernel.h:
```diff
@@ -6224,6 +6224,17 @@ int k_thread_runtime_stats_enable(k_tid_t thread);
+/**
+ * @brief Check if runtime statistics gathering is enabled for a thread
+ *
+ * This function checks the status of runtime statistics gathering for the
+ * specified thread and returns the current usage_lock status.
```
This makes no sense.

cfriedt replied:
Good catch

rruuaanng replied:
Do I need to remove it?

cfriedt replied:
I think the wording just needs to be changed.

rruuaanng replied:
How to change to `Checks whether runtime statistics collection is enabled for the specified thread`?

peter-mitsis replied:
@rruuaanng - I like the proprosed change in your comment--it is very clear.

stephanosio reviewed (changes requested):
Apart from the comments below, the commits should be reordered such that `kernel: Add k_thread_runtime_stats_is_enabled function` comes before `posix: Add posix CLOCK_THREAD_CPUTIME_ID support`.

cfriedt reviewed (commented):
Forgot I had one pending comment 😬

rruuaanng commented:
This PR has been deferred to 4.1. Maybe I won't be in such a hurry to handle it. I will prepare a PR for the implementation of `DT_FOREACH_ANCESTOR` today. Stay tuned ：）

rruuaanng commented:
I've change them, please review again :)
@cfriedt @stephanosio

rruuaanng commented:
CI throws. Emmm, I don't know :(
```
dts.cmake

warning: STDOUT_CONSOLE (defined at lib/libc/Kconfig:135) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: CONSOLE_HAS_DRIVER (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_STDOUT_CONSOLE and/or look up
STDOUT_CONSOLE in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

Parsing /__w/zephyr/zephyr/Kconfig
Loaded configuration '/__w/zephyr/zephyr/boards/m5stack/m5stack_cores3/m5stack_cores3_appcpu_defconfig'
Merged configuration '/__w/zephyr/zephyr/samples/synchronization/prj.conf'
Merged configuration '/__w/zephyr/zephyr/twister-out/m5stack_cores3_esp32s3_appcpu/samples/synchronization/sample.kernel.synchronization/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to '/__w/zephyr/zephyr/twister-out/m5stack_cores3_esp32s3_appcpu/samples/synchronization/sample.kernel.synchronization/zephyr/.config'
Kconfig header saved to '/__w/zephyr/zephyr/twister-out/m5stack_cores3_esp32s3_appcpu/samples/synchronization/sample.kernel.synchronization/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: /opt/toolchains/zephyr-sdk-0.17.0/xtensa-espressif_esp32s3_zephyr-elf/xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/toolchains/zephyr-sdk-0.17.0/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc
-- Espressif HAL path: /__w/zephyr/modules/hal/espressif
CMake Error at /__w/zephyr/zephyr/cmake/modules/extensions.cmake:4178 (message):
  dt_reg_addr(image_off ...) missing required argument: PATH
Call Stack (most recent call first):
  /__w/zephyr/zephyr/soc/espressif/esp32s3/CMakeLists.txt:78 (dt_reg_addr)

```

cfriedt commented:
@rruuaanng - this seems like an unrelated issue in CI. You might need to rebase once it's resolved on main.

ycsin commented on include/zephyr/kernel.h:
```diff
@@ -6224,6 +6224,17 @@ int k_thread_runtime_stats_enable(k_tid_t thread);
+ * statistics.
+ *
+ * @param thread ID of thread
+ * @return true if usage statistics are enabled, for the given thread, otherwise false
```
```suggestion
 * @return true if usage statistics are enabled for the given thread, otherwise false
```

ycsin reviewed (dismissed):
refresh +1 while I'm here before the CI passes

ycsin commented on tests/kernel/usage/thread_runtime_stats/src/test_thread_runtime_stats.c:
```diff
@@ -212,7 +212,7 @@ ZTEST(usage_api, test_thread_stats_enable_disable)
 	k_thread_runtime_stats_get(tid, &helper_stats1);
 	k_thread_runtime_stats_disable(tid);
-
+	zassert_false(k_thread_runtime_stats_is_enabled(tid));
```
Should have kept the empty line
```suggestion
	zassert_false(k_thread_runtime_stats_is_enabled(tid));

```

cfriedt commented:
Probably needs a rebate. qemu_arc_em was doing something kind of wacky, so I made a pr to disable it for the test that's failing.

andyross commented on kernel/usage.c:
```diff
@@ -272,6 +272,27 @@ int k_thread_runtime_stats_disable(k_tid_t  thread)
+
+	key = k_spin_lock(&usage_lock);
+	status = thread->base.usage.track_usage;
+	k_spin_unlock(&usage_lock, key);
```
Driving by: this lock is needless, remove.  You'd need to lock it if you were inspecting it and taking some action based on the value and want to be sure it's not changing while you make that decision.  But you're not doing anything but reading the value and returning it to an unsynchronized context.

rruuaanng replied:
I've change it.

rruuaanng commented:
Ehm... There are really many errors in the unittest
```
RunID: f2a3ca8e6d3fa1c0d4a920a1812b95e8
PROJECT EXECUTION FAILED

INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86_64_atom/tests/kernel/smp_boot_delay/kernel.multiprocessing.smp_boot_delay.minimallibc/handler.log
INFO    - 4 Iteration:
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue

INFO    - 698/698 qemu_x86_64/atom          kernel.multiprocessing.smp_boot_delay.minimallibc  FAILED Testsuite failed (qemu 1.354s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_x86_64_atom/tests/kernel/smp_boot_delay/kernel.multiprocessing.smp_boot_delay.minimallibc/handler.log
ERROR   - SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build v4.0.0-1254-gdce4594f9040 ***
Running TESTSUITE smp_boot_delay
```

rruuaanng commented:
@cfriedt This PR has been delayed for too long due to CI :(, but now, so far so good!

cfriedt commented:
@rruuaanng - needs a rebase again to resolve the merge conflict, but it would also be a good opportunity to add @ycsin's change request as well.

andyross commented on kernel/usage.c:
```diff
@@ -272,6 +272,20 @@ int k_thread_runtime_stats_disable(k_tid_t  thread)
+	CHECKIF(thread == NULL) {
+		errno = EINVAL;
+		return false;
+	}
```
Driving by again, but this nitpick rises to a should-fix level for me:

First, this is just bad error handling.  If a kernel API can fail at runtime, then it must return a signed integer and reserve the negative space of the result for error values (and document them!).  If a function returns bool (or any other by-value type), the assumption is that it's a predicate and has no failure conditions.

Second, this is needless defensive coding, which we don't do in Zephyr (especially in the kernel) as a matter of general practice.  Checking that an argument is valid is the job of the caller, not the callee.   If there is a particular usage mistake you're worried about (and "oops it was null" doesn't count, IMHO), then an __ASSERT is fine.

Just remove these lines, basically.

rruuaanng replied:
So, like this. Is this what you mean?
```
bool k_thread_runtime_stats_is_enabled(k_tid_t thread)
{
	__ASSERT(thread != NULL, "'thread' is not allowed to be NULL");

	return thread->base.usage.track_usage;
}
```

rruuaanng replied:
Throwing a message is more helpful for debugging than implicitly crashing. So I choose to do this :0

andyross replied:
I'd honestly prefer that check just be removed as it's not checking a particularly notable condition.  But it doesn't rise to a -1 and we can always remove it later on an "assertion diet" pass, which is probably needed anyway.

Consider: if every function in kernel.h that took a pointer had an assert like this, then CONFIG_ASSERT would be basically unusable on all the real platforms that need to use it!  This isn't Linux here, we need to run on devices with sub-megabyte space constraints.

rruuaanng replied:
OK, I will remove that assert.

andyross reviewed (changes requested):
One more nitpick that I can't get past.  But that's my only complaint, feel free to assume +1 once resolved.

rruuaanng commented on tests/posix/common/src/clock.c:
It no exists.
```
errorput@RUANG MINGW64 /e/code/github/zephyr/zephyr-main (gh80363)
$ cd tests/posix/timers/src/
bash: cd: tests/posix/timers/src/: No such file or directory
```

rruuaanng commented:
Please review them again! @cfriedt @ycsin @andyross

cfriedt commented on include/zephyr/kernel.h:
```diff
@@ -6418,6 +6418,17 @@ int k_thread_runtime_stats_enable(k_tid_t thread);
+/**
+ * @brief Check if runtime statistics gathering is enabled for a thread
+ *
+ * This routine checks whether the specified thread has enabled the of runtime
```
This should probably say "has enabled runtime statistics"

rruuaanng replied:
Done.

ycsin reviewed (dismissed):
🥳

ycsin commented:
The last 2 commits should be in another branch (and create a separate PR for them?)

rruuaanng commented:
> The last 2 commits should be in another branch (and create a separate PR for them?)

Oh, of course, I'll move them to the new PR.

rruuaanng commented:
```
INFO: Error found: portability.posix.c_lib_ext on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
INFO: Error found: portability.posix.c_lib_ext.minimal on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
INFO: Error found: portability.posix.c_lib_ext.newlib on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
INFO: Error found: portability.posix.c_lib_ext.picolibc on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
```
Hmmmm...

cfriedt commented:
> ```
> INFO: Error found: portability.posix.c_lib_ext on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> INFO: Error found: portability.posix.c_lib_ext.minimal on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> INFO: Error found: portability.posix.c_lib_ext.newlib on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> INFO: Error found: portability.posix.c_lib_ext.picolibc on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> ```
> Hmmmm...

Ugh... This is exactly why I thought #86634 was a mistake.

rruuaanng commented:
> > ```
> > INFO: Error found: portability.posix.c_lib_ext on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> > INFO: Error found: portability.posix.c_lib_ext.minimal on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> > INFO: Error found: portability.posix.c_lib_ext.newlib on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> > INFO: Error found: portability.posix.c_lib_ext.picolibc on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hmmmm...
> 
> Ugh... This is exactly why I thought #86634 was a mistake.

 Fine. It seems difficult to handle.

cfriedt commented:
@rruuaanng - it's possible that this has been fixed already. But you will need to rebase to fix merge conflicts anyway.

rruuaanng commented:
It seems to still exist :(
```
INFO: Total tests gathered: 9506
INFO: Error found: portability.posix.c_lib_ext on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
INFO: Error found: portability.posix.c_lib_ext.minimal on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
INFO: Error found: portability.posix.c_lib_ext.newlib on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
INFO: Error found: portability.posix.c_lib_ext.picolibc on qemu_cortex_m0/nrf51822 (Not enough RAM but is one of the integration platforms)
```

other:

https://github.com/zephyrproject-rtos/zephyr/issues/87561

ycsin commented:
@rruuaanng this needs another rebase

andyross commented on include/zephyr/kernel.h:
```diff
@@ -6576,6 +6576,16 @@ int k_thread_runtime_stats_enable(k_tid_t thread);
+ * @param thread ID of thread
+ * @return true if usage statistics are enabled for the given thread, otherwise false
+ */
+bool k_thread_runtime_stats_is_enabled(k_tid_t thread);
```
If this just depends on the thread struct, that header appears in the include paths already.  You can make this inline without worry and save a few code bytes for users.

andyross reviewed (dismissed):
Tiny nitpick.  I'll let others weigh in on the posix changes, seems like that could be a separate PR?  +1 for the meat of the PR.

cfriedt commented:
@rruuaanng - are you able to rebase?

rruuaanng committed:
kernel: Add k_thread_runtime_stats_is_enabled function

Add 'k_thread_runtime_stats_is_enabled' function, whichs
used to check whether runtime statistics collection is
enabled for a thread.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
tests: kernel: Add k_thread_runtime_stats_is_enabled function unittest

Add test for 'k_thread_runtime_stats_is_enabled(tid)'.

Signed-off-by: James Roy <rruuaanng@outlook.com>

cfriedt commented:
@rruuaanng - please rebase and resolve conflicts

rruuaanng commented:
@cfriedt @ycsin @peter-mitsis Please review again :)

cfriedt commented on kernel/usage.c:
```diff
@@ -271,6 +271,11 @@ int k_thread_runtime_stats_disable(k_tid_t  thread)
 	return 0;
 }
+
+bool k_thread_runtime_stats_is_enabled(k_tid_t thread)
```
Does this need to be a `__syscall`?

nashif replied:
dont think so, lets not overcomplicate it :)
